File: client_python/cliente.py
import grpc
import grpc
import sys
import os
from typing import Iterator

# Import gerados pelo protoc (assumidos em config_python)
from config_python import file_processor_pb2 as pb2
from config_python import file_processor_pb2_grpc as pb2_grpc

STORAGE_DIR = os.path.join(os.path.dirname(__file__), 'storage')


# Lista arquivos na pasta de storage
def list_storage_files() -> list[str]:
    if not os.path.isdir(STORAGE_DIR):
        os.makedirs(STORAGE_DIR, exist_ok=True)
    return [f for f in os.listdir(STORAGE_DIR) if os.path.isfile(os.path.join(STORAGE_DIR, f))]

# Permite ao usuário escolher um arquivo da pasta de storage
def choose_file() -> str:
    files = list_storage_files() # lista arquivos na pasta storage
    if not files:
        print(f"Pasta de storage vazia: {STORAGE_DIR}")
        return ''
    
    print("Arquivos disponíveis na pasta storage:")
    for i, name in enumerate(files, 1): # Itera sobre os arquivos disponíveis
        print(f"  {i}) {name}")

    while True:
        sel = input("Selecione o número do arquivo: ")
        try:
            idx = int(sel)
            if 1 <= idx <= len(files):
                return os.path.join(STORAGE_DIR, files[idx - 1])
        except ValueError:
            pass
        print("Seleção inválida. Tente novamente.")

# Gera stream de FileRequest a partir do arquivo e preenche os parâmetros
def stream_file_requests(path: str, params_filler) -> Iterator[pb2.FileRequest]:
    file_name = os.path.basename(path)
    first = True

    # Abre o arquivo e lê em chunks pouco a pouco para envio
    with open(path, 'rb') as f:
        while True:
            data = f.read(1024 * 1024)

            if not data:
                break

            # Cria chunk para requisição de envio
            chunk = pb2.FileChunk(content=data)
            req = pb2.FileRequest(file_name=file_name, file_content=chunk) # envia o chunk

            # Preenche os parâmetros específicos da operação na primeira requisição
            if first:
                params_filler(req)
                first = False
            yield req # Envia a requisição

# Grava as respostas do servidor em um arquivo de saída
def write_responses_to_file(responses, output_path: str):
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    with open(output_path, 'wb') as out:
        # Itera as respostas do servidor
        for resp in responses:
            if resp.file_content and resp.file_content.content:
                out.write(resp.file_content.content) # grava o conteúdo recebido
            if resp.status_message:
                print(f"[server] {resp.status_message} (success={resp.success})") # mostra mensagens de status

# Serviço de Compressão de PDF
def do_compress_pdf(stub, input_path: str):
    def fill_params(req: pb2.FileRequest): # Preenche parâmetros específicos
        req.compress_pdf_params.CopyFrom(pb2.CompressPDFRequest())

    # Chama o serviço do servidor
    responses = stub.CompressPDF(stream_file_requests(input_path, fill_params))

    # Define o caminho de saída
    base = os.path.splitext(os.path.basename(input_path))[0]
    output_path = os.path.join(STORAGE_DIR, f"{base}_compressed.pdf")
    
    # Escreve a resposta da requisição em arquivo de saída
    write_responses_to_file(responses, output_path)
    print(f"Saída salva em: {output_path}")


def do_convert_to_txt(stub, input_path: str):
    def fill_params(req: pb2.FileRequest): # Preenche parâmetros específicos
        req.convert_to_txt_params.CopyFrom(pb2.ConvertToTXTRequest())

    # Chama o serviço do servidor
    responses = stub.ConvertToTXT(stream_file_requests(input_path, fill_params))

    # Define o caminho de saída
    base = os.path.splitext(os.path.basename(input_path))[0]
    output_path = os.path.join(STORAGE_DIR, f"{base}.txt")

    # Escreve a resposta da requisição em arquivo de saída
    write_responses_to_file(responses, output_path)
    print(f"Saída salva em: {output_path}")


def do_convert_image_format(stub, input_path: str):
    # Solicita o formato de saída ao usuário
    out_format = input("Formato de saída (ex: png, jpg, webp): ").strip()

    def fill_params(req: pb2.FileRequest): # Preenche parâmetros específicos
        req.convert_image_format_params.CopyFrom(pb2.ConvertImageFormatRequest(output_format=out_format))

    # Chama o serviço do servidor
    responses = stub.ConvertImageFormat(stream_file_requests(input_path, fill_params))

    # Define o caminho de saída
    base = os.path.splitext(os.path.basename(input_path))[0]
    output_path = os.path.join(STORAGE_DIR, f"{base}.{out_format}")

    # Escreve a resposta da requisição em arquivo de saída
    write_responses_to_file(responses, output_path)
    print(f"Saída salva em: {output_path}")


def do_resize_image(stub, input_path: str):
    while True:
        try: # Solicita dimensões ao usuário
            width = int(input("Largura (px): ").strip())
            height = int(input("Altura (px): ").strip())
            break
        except ValueError:
            print("Valores inválidos, tente novamente.")

    # Preenche parâmetros específicos
    def fill_params(req: pb2.FileRequest):
        req.resize_image_params.CopyFrom(pb2.ResizeImageRequest(width=width, height=height))

    # Chama o serviço do servidor
    responses = stub.ResizeImage(stream_file_requests(input_path, fill_params))

    # Define o caminho de saída
    base = os.path.splitext(os.path.basename(input_path))[0]
    output_path = os.path.join(STORAGE_DIR, f"{base}_{width}x{height}.img")

    # Escreve a resposta da requisição em arquivo de saída
    write_responses_to_file(responses, output_path)
    print(f"Saída salva em: {output_path}")


def main():
    host = os.environ.get('GRPC_HOST', 'localhost')
    port = os.environ.get('GRPC_PORT', '50051')
    address = f"{host}:{port}"

    # Conecta ao servidor gRPC
    with grpc.insecure_channel(address) as channel:
        stub = pb2_grpc.FileProcessorServiceStub(channel)

        # Apresenta menu para seleção de serviços
        while True:
            print("\n=== Cliente Python ===")
            print("1) CompressPDF")
            print("2) ConvertToTXT")
            print("3) ConvertImageFormat")
            print("4) ResizeImage")
            print("0) Sair")

            opt = input("Escolha: ").strip()

            if opt == '0':
                break
            if opt not in {'1','2','3','4'}:
                print("Opção inválida")
                continue
            path = choose_file()
            if not path:
                continue

            # Chama o serviço de acordo com a opção escolhida
            try:
                if opt == '1':
                    do_compress_pdf(stub, path)
                elif opt == '2':
                    do_convert_to_txt(stub, path)
                elif opt == '3':
                    do_convert_image_format(stub, path)
                elif opt == '4':
                    do_resize_image(stub, path)
            # Erro na seleção do serviço
            except grpc.RpcError as e:
                print(f"Erro gRPC: {e.code()} - {e.details()}")


if __name__ == '__main__':
    main()

File: client_cpp/cliente.cpp
/*
 * Cliente gRPC em C++ para processamento de arquivos.
 * Padrão de comentários: estilo ANSI-C.
 *
 * Funcionalidades:
 *  - Lista arquivos na pasta client_cpp/storage.
 *  - Envia arquivo ao servidor via streaming para os 4 serviços.
 *  - Recebe arquivos de saída e grava em client_cpp/storage.
 */

#include <iostream>
#include <memory>
#include <string>
#include <fstream>
#include <vector>
#include <filesystem>

#include <grpcpp/grpcpp.h>

#include "../config_cpp/file_processor.grpc.pb.h"
#include "../config_cpp/file_processor.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using grpc::ClientReaderWriter;

using file_processor::FileProcessorService;
using file_processor::FileRequest;
using file_processor::FileResponse;
using file_processor::FileChunk;
using file_processor::CompressPDFRequest;

namespace fs = std::filesystem;

// Obtém o caminho do diretório de armazenamento do cliente
static std::string StorageDir() {
    return (fs::path(__FILE__).parent_path() / "storage").string();
}

// Lista os arquivos na pasta de storage
static std::vector<std::string> ListStorageFiles() {
    std::vector<std::string> files;
    fs::path dir(StorageDir());

    if (!fs::exists(dir)) fs::create_directories(dir); // Cria o diretório se não existir

    for (auto &p : fs::directory_iterator(dir)) { // Percorre os arquivos no diretório
        // Adiciona o nome do arquivo à lista
        if (p.is_regular_file()) files.push_back(p.path().filename().string()); 
    }

    return files;
}

// Retorna o diretório completo de um dos arquivos selecionados pelo usuário
static std::string ChooseFile() {
    auto files = ListStorageFiles(); // Lista arquivos

    if (files.empty()) {
        std::cout << "Pasta de storage vazia: " << StorageDir() << std::endl;
        return {};
    }

    //Imprime os arquivos disponíveis na pasta
    std::cout << "Arquivos disponíveis na pasta storage:" << std::endl;
    for (size_t i = 0; i < files.size(); ++i) std::cout << "  " << (i+1) << ") " << files[i] << std::endl;

    // Permite a seleção de um dos arquivos
    while (true) {
        std::cout << "Selecione o número do arquivo: ";
        int idx;

        if (!(std::cin >> idx)) { 
            std::cin.clear(); std::cin.ignore(1024, '\n'); 
            continue; 
        }

        // Se número selecionado for válido, retorna o caminho completo
        if (idx >= 1 && (size_t)idx <= files.size()) return (fs::path(StorageDir()) / files[idx-1]).string();

        // Se não, continua o loop
        std::cout << "Seleção inválida. Tente novamente." << std::endl;
    }
}

// Cliente gRPC para FileProcessorService
class FileProcessorClient {
public:
    // Cria stub gRPC para comunicação com o servidor
    FileProcessorClient(std::shared_ptr<Channel> channel): 
        stub_(FileProcessorService::NewStub(channel)) {}

    // Serviço de compressão de PDF
    bool CompressPDF(const std::string& input_path, const std::string& output_path) {
        // Cria contexto gRPC
        ClientContext context;

        // Inicia chamada GRPC bidirecional para o método CompressPDF
        std::unique_ptr<ClientReaderWriter<FileRequest, FileResponse>> stream(stub_->CompressPDF(&context));

        // Cria fluxo de leitura do arquivo de entrada
        std::ifstream in(input_path, std::ios::binary);

        // Verifica se o arquivo foi aberto corretamente
        if (!in) {
            std::cerr << "Falha ao abrir arquivo de entrada: " << input_path << std::endl;
            return false;
        }

        // Enviar parâmetros no primeiro chunk, depois dados em pedaços
        {
            FileRequest req;
            req.set_file_name(fs::path(input_path).filename().string());
            req.mutable_compress_pdf_params();
            stream->Write(req);
        }

        // Envia o arquivo em pedaços
        {
            const size_t CHUNK = 1024 * 1024;
            std::vector<char> buf(CHUNK);
            while (in) {
                // Lê parte pré determinada pelo tamanho do chunk
                in.read(buf.data(), buf.size());
                std::streamsize n = in.gcount(); // Tamanho do stream

                if (n <= 0) break;

                FileRequest req; // Cria nova requisição para envio
                req.set_file_name(fs::path(input_path).filename().string());
                req.mutable_file_content()->set_content(buf.data(), static_cast<size_t>(n));

                // Envia o chunk para o servidor
                if (!stream->Write(req)) break;
            }
        }
        stream->WritesDone();

        // Recebe as respostas do servidor e grava no arquivo de saída
        FileResponse resp;
        std::ofstream out(output_path, std::ios::binary);
        while (stream->Read(&resp)) { // Enquanto houver conteúdo para leitura no arquivo recebido
            if (resp.has_file_content()) { // Se tiver conteúdo, escrever no arquivo de saída
                out.write(resp.file_content().content().data(), resp.file_content().content().size());
            }
            // Exibe status da operação
            std::cout << "[server] success=" << resp.success() << " message=" << resp.status_message() << std::endl;
        }

        // Finaliza a chamada gRPC e verifica o status
        Status status = stream->Finish();
        if (!status.ok()) {
            std::cerr << "gRPC failed: " << status.error_message() << std::endl;
            return false;
        }
        return true;
    }

    bool ConvertToTXT(const std::string& input_path, const std::string& output_path) {
        // Cria contexto gRPC
        ClientContext context;
        // Inicia chamada GRPC bidirecional para o método ConvertToTXT
        auto stream = stub_->ConvertToTXT(&context);
        {
            // Enviar parâmetros no primeiro chunk, depois dados em pedaços
            FileRequest req; req.set_file_name(fs::path(input_path).filename().string());
            req.mutable_convert_to_txt_params();
            stream->Write(req);
        }

        // Cria fluxo de leitura do arquivo de entrada
        std::ifstream in(input_path, std::ios::binary);
        const size_t CHUNK = 1024*1024; std::vector<char> buf(CHUNK);

        // Envia o arquivo em pedaços
        while (in) { 
            in.read(buf.data(), buf.size()); 
            auto n=in.gcount();

            if (n<=0) break;

            FileRequest req; 
            req.set_file_name(fs::path(input_path).filename().string()); 
            req.mutable_file_content()->set_content(buf.data(), (size_t)n);

            if(!stream->Write(req)) break; 
        }

        stream->WritesDone();

        // Recebe as respostas do servidor e grava no arquivo de saída
        std::ofstream out(output_path, std::ios::binary); FileResponse resp;
        while (stream->Read(&resp)) 
            if (resp.has_file_content()) 
                out.write(resp.file_content().content().data(), resp.file_content().content().size());
        auto status = stream->Finish(); if(!status.ok()){ std::cerr<<"gRPC failed: "<<status.error_message()<<std::endl; return false; }
        return true;
    }

    bool ConvertImageFormat(const std::string& input_path, const std::string& output_path, const std::string& format) {
        ClientContext context; 
        auto stream = stub_->ConvertImageFormat(&context);
        { 
            FileRequest req; 
            req.set_file_name(fs::path(input_path).filename().string()); 
            req.mutable_convert_image_format_params()->set_output_format(format); 
            stream->Write(req);
        }

            std::ifstream in(input_path, std::ios::binary); 
            const size_t CHUNK=1024*1024; std::vector<char> buf(CHUNK);
            
            while(in){ 
                in.read(buf.data(), buf.size()); auto n=in.gcount();

                if(n<=0) break;

                FileRequest req; 
                req.set_file_name(fs::path(input_path).filename().string()); 
                req.mutable_file_content()->set_content(buf.data(), (size_t)n); 
                if(!stream->Write(req)) break; 
            }

            stream->WritesDone(); std::ofstream out(output_path, std::ios::binary); 
            FileResponse resp;

            while(stream->Read(&resp)) 
                if(resp.has_file_content()) 
                    out.write(resp.file_content().content().data(), resp.file_content().content().size());
            
            auto status=stream->Finish(); 
            
            if(!status.ok()){ 
                std::cerr<<"gRPC failed: "<<status.error_message()<<std::endl; 
                return false;
            } 
            
            return true;
    }

    bool ResizeImage(const std::string& input_path, const std::string& output_path, int width, int height) {
        // Cria contexto gRPC
        ClientContext context; 
        auto stream = stub_->ResizeImage(&context);

        // Envia parâmetros da requisição
        { 
            FileRequest req; 
            req.set_file_name(fs::path(input_path).filename().string()); 
            auto* p=req.mutable_resize_image_params(); 
            p->set_width(width); p->set_height(height); 
            stream->Write(req);
        }

            // Cria fluxo de leitura do arquivo de entrada
            std::ifstream in(input_path, std::ios::binary); 
            const size_t CHUNK=1024*1024; 
            std::vector<char> buf(CHUNK);

            // Envia o arquivo em pedaços
            while(in){ 
                in.read(buf.data(), buf.size()); 
                auto n=in.gcount(); 

                if(n<=0) break; 

                FileRequest req; 
                req.set_file_name(fs::path(input_path).filename().string()); 
                req.mutable_file_content()->set_content(buf.data(), (size_t)n); 

                if(!stream->Write(req)) break; 
            }

            // Indica que todos os dados foram enviados
            stream->WritesDone(); 

            // Cria fluxo de escrita do arquivo de saída
            // Passa o que foi recebido do servidor ao diretório de saída
            std::ofstream out(output_path, std::ios::binary); 
            FileResponse resp;

            // Recebe as respostas do servidor e grava no arquivo de saída
            while(stream->Read(&resp)) 
                if(resp.has_file_content()) 
                    out.write(resp.file_content().content().data(), resp.file_content().content().size());

            auto status=stream->Finish(); 

            if(!status.ok()){ 
                std::cerr<<"gRPC failed: "<<status.error_message()<<std::endl; 
                return false;
            } 

            return true;
    }

private:
    // Stub gRPC para comunicação com o servidor
    std::unique_ptr<FileProcessorService::Stub> stub_;
};

int main() {
    std::string server = "localhost:50051";

    // Cria canal de comunicação com o servidor
    FileProcessorClient client(grpc::CreateChannel(server, grpc::InsecureChannelCredentials()));

    // Criação de Menu para seleção dos serviços
    while (true) {
        std::cout << "\n=== Cliente C++ ===\n1) CompressPDF\n2) ConvertToTXT\n3) ConvertImageFormat\n4) ResizeImage\n0) Sair\nEscolha: ";

        int opt; 
        
        if(!(std::cin>>opt)){ 
            std::cin.clear(); 
            std::cin.ignore(1024,'\n'); 
            continue; 
        }

        if (opt==0) break;

        std::cin.ignore(1024,'\n');

        std::string input_path = ChooseFile();

        if (input_path.empty()) continue;
        fs::path in(input_path); std::string base = in.stem().string();

        if (opt==1) {
            std::string out = (fs::path(StorageDir()) / (base+"_compressed.pdf")).string();
            client.CompressPDF(input_path, out);
            std::cout << "Saída: " << out << std::endl;
        } else if (opt==2) {
            std::string out = (fs::path(StorageDir()) / (base+".txt")).string();
            client.ConvertToTXT(input_path, out);
            std::cout << "Saída: " << out << std::endl;
        } else if (opt==3) {
            std::string format; std::cout << "Formato de saída (png/jpg/webp): "; std::getline(std::cin, format);
            std::string out = (fs::path(StorageDir()) / (base+"."+format)).string();
            client.ConvertImageFormat(input_path, out, format);
            std::cout << "Saída: " << out << std::endl;
        } else if (opt==4) {
            int w,h; std::cout << "Largura: "; std::cin>>w; std::cout << " Altura: "; std::cin>>h; std::cin.ignore(1024,'\n');
            std::string out = (fs::path(StorageDir()) / (base+"_"+std::to_string(w)+"x"+std::to_string(h)+".img")).string();
            client.ResizeImage(input_path, out, w, h);
            std::cout << "Saída: " << out << std::endl;
        }
    }
    return 0;
}
